### ISSUE TITLE
tests: dma: chan_blen_transfer: test breakage on Nucleo-H743ZI

### ISSUE THREAD

**Issue by mathieuchopstm** [zephyrproject-rtos/zephyr#75676]
**Describe the bug**

(follow-up issue to #75125)

After merge of #75245, the `chan_blen_transfer` test now succeeds on STM32WB0 MCUs but fails on Nucleo-H743ZI.

Relevant test log extract:
```
===================================================================                                                                       
START - test_tst_dma1_m2m_chan0_burst16                                                                                                   
Preparing DMA Controller: Name=dmamux@58025800, Chan_ID=4, BURST_LEN=2                                                                    
Starting the transfer                                                                                                                     
E: invalid source address                                                                                                                 
E: cannot configure the dmamux. 
```

The `invalid source address` message is emitted here:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/drivers/dma/dma_stm32_bdma.c#L566-L573
where `bdma_stm32_is_valid_memory_address` is:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/drivers/dma/dma_stm32_bdma.c#L462-L478

TL;DR: **the DMA driver refuses the transfer because the source is not in SRAM4**.

The regression comes from the fact that, beforehand, the `.rodata` and `.bss` segments of the test code could be relocated thanks to:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/tests/drivers/dma/chan_blen_transfer/CMakeLists.txt#L10

And indeed, the Kconfig for Nucleo-H743ZI makes use of this feature:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/tests/drivers/dma/chan_blen_transfer/boards/nucleo_h743zi.conf#L5-L9

Before #75245, the test looked like this:
https://github.com/zephyrproject-rtos/zephyr/blob/4180d70439f599a3597b2e31753773cfa222ace2/tests/drivers/dma/chan_blen_transfer/src/test_dma.c#L25-L26
 - `tx_buff` is `static const` and initialized, so it goes in `.rodata` and gets relocated
 - `rx_buff` is `static`, not `const`, but uninitialized, so it goes in `.bss` and gets relocated 
After the patch, it now looks like this:
https://github.com/zephyrproject-rtos/zephyr/blob/662b9803fa9985f499d3627f96589a1ed8d2c8dc/tests/drivers/dma/chan_blen_transfer/src/test_dma.c#L25-L26
 - `tx_buff` is `static`, not `const`, and initialized, so it goes in `.data` **and is not relocated**
 - `rx_buff` is `static`, not `const`, but uninitialized, so it goes in `.bss` and gets relocated 

**Expected behavior**
Test should not fail.

**Impact**
Test breakage on Nucleo-H743ZI and potentially all boards with a Kconfig overlay setting `CONFIG_DMA_LOOP_TRANSFER_RELOCATE_SECTION` (i.e., `mimxrt` boards).

**Comment by mathieuchopstm** [comment]
Potential fix proposition: move `tx_data` and `rx_data` to a separate file, and relocate the whole file.

Appears to work on STM32WB0, under testing on STM32H7... patch would look something like this:
```diff
diff --git a/tests/drivers/dma/chan_blen_transfer/CMakeLists.txt b/tests/drivers/dma/chan_blen_transfer/CMakeLists.txt
--- a/tests/drivers/dma/chan_blen_transfer/CMakeLists.txt
+++ b/tests/drivers/dma/chan_blen_transfer/CMakeLists.txt
@@ -7,4 +7,6 @@ project(chan_blen_transfer)
 FILE(GLOB app_sources src/*.c)
 target_sources(app PRIVATE ${app_sources})
 
-zephyr_code_relocate(FILES src/test_dma.c LOCATION ${CONFIG_DMA_LOOP_TRANSFER_RELOCATE_SECTION}_RODATA_BSS)
+if (CONFIG_DMA_LOOP_TRANSFER_RELOCATE_SECTION)
+  zephyr_code_relocate(FILES src/test_dma_data.c LOCATION ${CONFIG_DMA_LOOP_TRANSFER_RELOCATE_SECTION})
+endif()
diff --git a/tests/drivers/dma/chan_blen_transfer/src/test_dma.c b/tests/drivers/dma/chan_blen_transfer/src/test_dma.c
--- a/tests/drivers/dma/chan_blen_transfer/src/test_dma.c
+++ b/tests/drivers/dma/chan_blen_transfer/src/test_dma.c
@@ -20,10 +20,10 @@
 #include <zephyr/drivers/dma.h>
 #include <zephyr/ztest.h>
 
-#define RX_BUFF_SIZE (48)
+#include "shared.h"
 
-static __aligned(32) const char tx_data[] = "It is harder to be kind than to be wise........";
-static __aligned(32) char rx_data[RX_BUFF_SIZE] = { 0 };
+extern char tx_data[DATA_BUF_SIZE];
+extern char rx_data[DATA_BUF_SIZE];
 
 static void test_done(const struct device *dma_dev, void *arg,
 		      uint32_t id, int status)
diff --git a/tests/drivers/dma/chan_blen_transfer/src/shared.h b/tests/drivers/dma/chan_blen_transfer/src/shared.h
new file mode 100644
--- /dev/null
+++ b/tests/drivers/dma/chan_blen_transfer/src/shared.h
@@ -0,0 +1,3 @@
+
+#define DATA_BUF_SIZE (48)
+
diff --git a/tests/drivers/dma/chan_blen_transfer/src/test_dma_data.c b/tests/drivers/dma/chan_blen_transfer/src/test_dma_data.c
new file mode 100644
--- /dev/null
+++ b/tests/drivers/dma/chan_blen_transfer/src/test_dma_data.c
@@ -0,0 +1,7 @@
+
+#include <zephyr/kernel.h>
+
+#include "shared.h"
+
+__aligned(32) char tx_data[DATA_BUF_SIZE] = "It is harder to be kind than to be wise........";
+__aligned(32) char rx_data[DATA_BUF_SIZE] = { 0 };
```

**Comment by mathieuchopstm** [comment]
Fix proposed in #75895.
